File: client/src/lib/pdf-utils.ts
import { Document } from "@shared/schema";

export interface ProcessedPDFContent {
  text: string;
  pageCount: number;
  title?: string;
  author?: string;
  creationDate?: string;
}

// Convert base64 to raw binary data
export function base64ToArrayBuffer(base64: string): ArrayBuffer {
  const binaryString = window.atob(base64);
  const len = binaryString.length;
  const bytes = new Uint8Array(len);
  
  for (let i = 0; i < len; i++) {
    bytes[i] = binaryString.charCodeAt(i);
  }
  
  return bytes.buffer;
}

// Extract text content from PDF using PDF.js
export async function extractPDFContent(base64Content: string): Promise<ProcessedPDFContent> {
  try {
    // This would normally use PDF.js to extract text
    // For demo purposes, we'll simulate the extraction
    return {
      text: "This is the extracted text content from the PDF document. It would normally contain the actual text extracted from all pages of the PDF file.",
      pageCount: 5,
      title: "Financial Document",
      author: "Financial Institution",
      creationDate: new Date().toISOString()
    };
  } catch (error) {
    console.error("Error extracting PDF content:", error);
    throw new Error("Failed to extract content from PDF");
  }
}

// Get the data URL for displaying a PDF from a document
export function getPDFDataUrl(document: Document): string {
  if (document.fileType !== "pdf") {
    throw new Error("Document is not a PDF");
  }
  
  return `data:application/pdf;base64,${document.fileContent}`;
}

// Helper to create a file download for a PDF document
export function downloadPDF(document: Document): void {
  if (document.fileType !== "pdf") {
    throw new Error("Document is not a PDF");
  }
  
  const linkSource = `data:application/pdf;base64,${document.fileContent}`;
  const downloadLink = document.createElement("a");
  
  downloadLink.href = linkSource;
  downloadLink.download = document.title;
  downloadLink.click();
}

File: python_api/app/utils/pdf_utils.py
import base64
import io
from typing import Dict, Any, Optional
from pypdf import PdfReader

def base64_to_bytes(base64_string: str) -> bytes:
    """Convert a base64 string to bytes."""
    # Remove potential data URL prefix
    if ',' in base64_string:
        base64_string = base64_string.split(',', 1)[1]
    return base64.b64decode(base64_string)

def extract_pdf_content(base64_content: str) -> Dict[str, Any]:
    """
    Extract text and metadata from a PDF file encoded as base64.
    
    Args:
        base64_content (str): The PDF content encoded as base64
        
    Returns:
        Dict containing extracted text, page count and metadata
    """
    pdf_bytes = base64_to_bytes(base64_content)
    pdf_file = io.BytesIO(pdf_bytes)
    
    try:
        reader = PdfReader(pdf_file)
        text = ""
        
        # Extract text from all pages
        for page in reader.pages:
            page_text = page.extract_text()
            if page_text:
                text += page_text + "\n\n"
        
        # Get metadata
        info = reader.metadata
        
        return {
            "text": text,
            "page_count": len(reader.pages),
            "title": info.title if info and info.title else None,
            "author": info.author if info and info.author else None,
            "creation_date": info.creation_date if info and info.creation_date else None,
        }
    except Exception as e:
        raise Exception(f"Failed to extract PDF content: {str(e)}")

def get_pdf_data_url(content_base64: str) -> str:
    """
    Get a data URL for a PDF file encoded as base64.
    
    Args:
        content_base64 (str): The PDF content encoded as base64
        
    Returns:
        str: Data URL for the PDF
    """
    # Check if it's already a data URL
    if content_base64.startswith('data:application/pdf;base64,'):
        return content_base64
    else:
        return f'data:application/pdf;base64,{content_base64}'